atlasdb-commons/src/main/java/com/palantir/util/IndexEncodingUtils.java

```diff
@@ -0,0 +1,203 @@
+                    UnsafeArg.of("unknownKeys", unknownKeys));
+        }
+        return IndexEncodingResult.<K, R>builder()
+                .keyList(keyList)
+                .indexToValue(indexToValue)
+                .keyListChecksum(computeChecksum(checksumType, keyList))
+                .build();
+    }
+
+    /**
+     * Compute a derived map, replacing indices into the list of keys with their item, to the value returned
+     * by running the {@code valueMapper} over the original value.
+     *
+     * @param indexEncoding the output of {@link IndexEncodingUtils#encode}, i.e. the ordered list of keys,
+     * a map of indices to values, and a checksum of the ordered key list. Every index must be
+     * a valid index into the list of keys
+     * @param valueMapper a mapping function applied to values before placing them into the result map
+     * @throws SafeIllegalArgumentException if the provided checksum does not match the actual checksum of the ordered
+     * keys or if {@code indexToValue} contains indices that are not valid for the list of keys
+     */
+    public static <K extends DeterministicHashable, V, R> Map<K, R> decode(
+            IndexEncodingResult<K, V> indexEncoding, Function<V, R> valueMapper) {
+        List<K> keyList = indexEncoding.keyList();
+        KeyListChecksum expectedChecksum = indexEncoding.keyListChecksum();
+        KeyListChecksum actualChecksum = computeChecksum(expectedChecksum.type(), keyList);
+        Preconditions.checkArgument(
+                actualChecksum.equals(expectedChecksum),
+                "Key list integrity check failed",
+                UnsafeArg.of("keyList", keyList),
+                UnsafeArg.of("actualChecksum", actualChecksum),
+                UnsafeArg.of("expectedChecksum", expectedChecksum));
+
+        Map<K, R> keyToValue =
+                Maps.newHashMapWithExpectedSize(indexEncoding.indexToValue().size());
+        for (Map.Entry<Integer, V> entry : indexEncoding.indexToValue().entrySet()) {
+            int index = entry.getKey();
+            Preconditions.checkArgument(
+                    index >= 0 && index < keyList.size(),
+                    "Index map contains invalid index",
+                    UnsafeArg.of("index", index),
+                    UnsafeArg.of("keyListSize", keyList.size()));
+            keyToValue.put(keyList.get(index), valueMapper.apply(entry.getValue()));
+        }
+        return keyToValue;
+    }
+
+    @VisibleForTesting
+    static <K extends DeterministicHashable> KeyListChecksum computeChecksum(
+            ChecksumType checksumType, List<K> keyList) {
+        byte[] checksumValue;
+        switch (checksumType) {
+            case CRC32_OF_DETERMINISTIC_HASHCODE: {
+                CRC32 checksum = new CRC32();
+                for (K key : keyList) {
+                    checksum.update(key.deterministicHashCode());
+                }
+                checksumValue =
```

Comment:
Let's just early return from here with KeyListChecksum, rather than having the variable declaration outside of the switch